src/Microsoft.SymbolStore.Client/SymbolLocator.cs

```diff
@@ -0,0 +1,169 @@
+            SymbolServerResult result = await SearchServers((ISymbolServer server) => server.FindPEFileAsync(filename, timestamp, filesize));
+            if (result == null)
+                return null;
+
+            // The result may be a cab file instead of the actual result.  In this case, decompress it.
+            if (!result.Compressed)
+                return Cache.StorePEFile(result.Stream, filename, timestamp, filesize);
+            
+            CabConverter converter = new CabConverter(result.Stream);
+            using (MemoryStream stream = await converter.ConvertAsync())
+                return Cache.StorePEFile(stream, filename, timestamp, filesize);
+        }
+
+        private async Task<SymbolServerResult> SearchServers(Func<ISymbolServer, Task<SymbolServerResult>> getTask)
+        {
+            IEnumerable<Task<SymbolServerResult>> tasks = _symbolServers.Where(server => server.PreferThisServer).Select(server => getTask(server));
+            Task<SymbolServerResult> primaryResult = GetFirstNonNullResult(new List<Task<SymbolServerResult>>(tasks));
+
+            tasks = _symbolServers.Where(server => !server.PreferThisServer).Select(server => getTask(server));
+            Task<SymbolServerResult> secondaryResult = GetFirstNonNullResult(new List<Task<SymbolServerResult>>(tasks));
+
+            return await primaryResult ?? await secondaryResult;
+        }
+
+        private async Task<SymbolServerResult> FindPdbWorkerAsync(string pdbName, Guid guid, int age)
+        {
+            List<Task<SymbolServerResult>> processing = new List<Task<SymbolServerResult>>(_symbolServers.Length);
+            List<ISymbolServer> onDeck = new List<ISymbolServer>(_symbolServers.Length);
+
+            Task<SymbolServerResult> preferredResult = SearchServers((ISymbolServer server) => server.PreferThisServer, (ISymbolServer server) => server.FindPdbAsync(pdbName, guid, age));
+            Task<SymbolServerResult> secondaryResult = SearchServers((ISymbolServer server) => !server.PreferThisServer, (ISymbolServer server) => server.FindPdbAsync(pdbName, guid, age));
+
+            SymbolServerResult result = await preferredResult;
+            if (result != null)
+                return result;
+
+            return await secondaryResult;
+        }
+
+        private Task<SymbolServerResult> SearchServers(Func<ISymbolServer, bool> predicate, Func<ISymbolServer, Task<SymbolServerResult>> getTask)
+        {
+            var tasks = _symbolServers.Where(server => predicate(server)).Select(server => getTask(server));
+            return GetFirstNonNullResult(new List<Task<SymbolServerResult>>(tasks));
+        }
+        
+        async Task<T> GetFirstNonNullResult<T>(List<Task<T>> tasks) where T : class
+        {
+            while (tasks.Count > 0)
+            {
+                Task<T> task = await Task.WhenAny(tasks);
+
+                T result = task.Result;
+                if (result != null)
+                    return result;
```

Comment:
>return result; [](start = 20, length = 14)

Not needed for checkin, but adding cancelation support might make some nice perf improvements in the future. As I understand the current algorithm we initiate a download from multiple servers. Even after one of them is finished we continue to download from all the other servers in the background and then allow their work to be GC'ed without using it.